### ISSUE TITLE
floating point exception in linalg / lapack with intel compiler

### ISSUE THREAD

**Issue by chem** [numpy/numpy#6999]
I've encountered an issue with underflows in numpy.  Example code as follows:
```
#!/usr/bin/env python
import numpy as np
import numpy.linalg as la

a=np.asarray([[0.70710678,0.],[0.70710678,1.]])
b=np.asarray([0.4, 0.2])
c=5.0e-16

print 'in'
r1, r2, r3, r4 = la.lstsq(a, b, c)
print 'out'
```
A floating point exception is thrown before 'out' is reached, when using intel compilers with the "-fp-trap-all=underflow" flag, and numpy 1.10.2 / python 2.7.11.  Compiles using gcc or intel and *not* using underflow detection will complete successfully.  I eventually tracked the offending line down to L1918 of linalg.py:
```
        results = lapack_routine(m, n, n_rhs, a, m, bstar, ldb, s, rcond,
                                 0, work, lwork, iwork, 0)
```

lapack_routine is making a call to lapack_lite.dgelsd -- so, a C library is being loaded for that call to a LAPACK function, and something during the call underflows.  Underflows are unusual traps; not set by default with gcc or intel's fp-trap-all=common, but I am interested in detecting them elsewhere should they occur.  It would be great if numpy could run with it enabled.

icc --version
icc (ICC) 15.0.3 20150407

System is a Linux machine, SGI ICE X running SLES11 and kernel 3.0.

I compiled numpy without linking intel MKL, since the offending lapack_lite call may be affected by it, and obviously we're interested in the non-MKL version on numpy's bug tracker.  To disable MKL being linked I unset my MKLROOT and disabled MKL detection in numpy/distutils/system_info.py.  ldd confirms that mkt_rt.so was not linked, nor any other blas/lapack library.  Otherwise I generally followed intel's instructions for compilation at: https://software.intel.com/en-us/articles/numpyscipy-with-intel-mkl , while adding -fp-trap-all=underflow to the lines in numpy/distutils/intelccompiler.py with xhost/fpic/etc.

I see that the lapack_lite code is a C translation of Fortran LAPACK, so I'm not sure if this issue is solvable on the numpy end.  The objective would be to have my original code sample run with underflow traps enabled using intel compilers.  Can numpy devs help?  If not, could they bump the issue up to the appropriate place (perhaps LAPACK itself)?

I also noticed that all the lapack code in dlapack_lite.c appears to be based on LAPACK version 3.0 (according to the code comments), which is now 16 years old.  LAPACK 3.6.0 was released in November, 2015.  Would numpy benefit from updating its LAPACK rountines?

Thank you to those who work on numpy; great package.

**Comment by rgommers** [comment]
@chem there's probably not much interest in updating ``lapack_lite`` to a newer version, since it's inherently limited. For decent performance a real BLAS/LAPACK should be linked - numpy supports all the major ones. 

If you want to compare MKL to something else, maybe just link to a recent reference blas/lapack (or ATLAS, or OpenBLAS)?

About the underflow issue, that doesn't look familiar. May be compiler/OS specific - both icc and SGI + SLES aren't very regularly tested.

**Comment by chem** [comment]
yeah, I was trying to avoid use of MKL, ATLAS, or OpenBLAS in this bug report because if there's an underflow in dgelsd during the testcase when using those, then the issue report should be made to those projects, not numpy.  The underflow not looking familiar is why I thought this would be worthy of a report.  :)   I duplicated this problem on another platform, icc/Cray/SLES.  I can't duplicate with gcc without modifying numpy because of the lack of a compiler flag to detect underflow in C/C++.

Aside from my pure numpy example, this problem is important because of the following use pattern:  1. compile python/numpy without floating point traps.  2. run a python script that loads both numpy and a different C library, which *has* been compiled with fp-trap-all and intel compilers.   When the numpy code is executed, the CPU floating point traps will be "active" due to the loading of the library and fp-trap-all, even if the call to numpy.linalg does not involve the different C library.  So, numpy will crash.

In scientific computing, that pattern is sometimes encountered in cases where it is important to have underflow detection enabled.  Workarounds are possible, but in a perfect world they would not be needed.

Some users (especially hobbyist ones, that are not on SGI or Cray) may not have a LAPACK library available so I think there would be value in keeping lapack_lite up to date in general.  I think it is of great value that a third-party LAPACK library is not needed for a local install of numpy, even if its performance will not compare to an optimized library.

Thanks for reading!
